wordpress-mobile/WordPress-Android
# MSD - Handle Single Dashboard Card Error in UI

Parent #15750
References [pbArwn-3ny-p2](pbArwn-3ny-p2) (`if one card fails - error inside card` state and title copy change in [this comment](https://github.com/wordpress-mobile/WordPress-Android/pull/15763#issuecomment-1000841278))
References https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2251

This PR handles errors within `Posts` and `Today's Stats` cards using mocked data.

Note: As `unauthorised`, `jetpack_disconnected` and `jetpack_disabled` errors are irrecoverable (see p1643789176707579-slack-C0290FLA0RM), these error are filtered out from displaying. The condition for showing these error cards can be modified in the future in the corresponding builder classes. The tests ensure that no card (neither with error nor with data) is displayed when these errors are returned for that card type.

To test:

Prerequisites

- Launch the app.
- Select a `wpcom` site.
- Navigate to `My Site` -> `App Settings` -> `Debug Settings`.
- Ensure that the `MySiteDashboardPhase2FeatureConfig` is to set to on.
- Restart the app.

1. Notice that no card exists for `Today's Stats` and `Posts` in the `My Site Dashboard` with the given [mocked JSON](https://github.com/wordpress-mobile/WordPress-Android/blob/1e5728110e64e11000bc792db582329d9ec61173/WordPress/src/main/res/raw/mocked_site_dashboard_cards.json).
2. [Optional, [covered in unit tests](https://github.com/wordpress-mobile/WordPress-Android/pull/15893/commits/414e40aafa802fb28a2473b17a1029707c93cd5c)] Replace mocked JSON `error` node with different supported error types and reinstall the app:
    ```
    Today's Stats: `unauthorized`, `jetpack_disconnected` and `jetpack_disabled` 
    Posts: `unauthorized`
    ```
    Notice that these errors are never shown.

3.  Replace mocked JSON `error` node with a fallback `generic_error` and reinstall the app. Notice that a generic error within a card is displayed as shown. We can further discuss showing fallback errors in this Slack thread (p1644303679654909/1643829887.600439-slack-C0290FLA0RM). An error within a card, when shown, is tracked as follows (see pbArwn-2zs-p2#comment-4731). Tracking can be tested by enabling `My Site` -> `App Settings` -> `Privacy settings` -> `Collect information` and switching between sites.

    ```
    🔵 Tracked: my_site_dashboard_card_shown, Properties: {"type":"error","subtype":"todays_stats"}
    🔵 Tracked: my_site_dashboard_card_shown, Properties: {"type":"error","subtype":"post"}
    ```
    <img width=320 src="https://user-images.githubusercontent.com/1405144/152790086-655070fe-b250-4530-ac9d-ec890f65d711.png"/>


**Merge Instructions**

This PR is still in draft mode as it depends on the previous unmerged `Today's Card` UI with mocked data and the linked draft `FluxC` PR requiring changes after the endpoint is live. Please wait to merge the PR until mentioned PRs are merged to `develop`.

## Regression Notes
1. Potential unintended areas of impact
Cards should not be displayed with default values (like 0 likes, 0 views) when an error exists. 

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested.

3. What automated tests I added (or what prevented me from doing so)
Add unit tests. 

PR submission checklist:
